### ISSUE TITLE
Add middleware to allow one service to redirect to another by a response header

### ISSUE THREAD

**Issue by zyp** [traefik/traefik#5154]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

From one service, I would like to be able to redirect to another service by returning a response header similar to this:

```
X-Redirect: http://foo-service/bar
```

Alternatively a list of services that can be redirected to can be configured in the middleware and the header can look like this:

```
X-Redirect: foo-service /bar
```

### Use case

We have an application that's currently using `X-Sendfile` to serve up big files without having to pass the data through the application itself. We're now in the process of switching from having the data stored on the local filesystem to storing it in S3-style object storage and would like an analogous way to tell Traefik to serve the file directly from the storage service.

I figure the implementation would look similar to the Errors middleware, except the redirect would be based on a custom response header rather than the response status.

**Comment by dduportal** [comment]
Hi @zyp ,

Is there a particular reason why the backend is not answering a standard `Location` HTTP header that would trigger the redirect at the client?

Also, please note that Traefik is not aimed at serving files (and does not :) ). So you'll need a specific backend service that would implement this logic right?

**Comment by zyp** [comment]
> Is there a particular reason why the backend is not answering a standard Location HTTP header that would trigger the redirect at the client?

Yes, I consider the storage server a detail of the internal infrastructure that the client doesn't need to know about. Secondly, the application performs a sort of forward auth, so I don't want clients to be able to bypass that. Thirdly, some of the files are only cached and if they don't exist they are generated on demand.

The ForwardAuth middleware would get me some of the way, but it's a bit too inflexible for my use. It doesn't allow me to rewrite the request URL and since a 200-response is interpreted as a proceed, it's not possible to return a successful response directly from the auth service.

> Also, please note that Traefik is not aimed at serving files (and does not :) ). So you'll need a specific backend service that would implement this logic right?

Yes. Like mentioned, I'll point it at a storage service with an S3-style interface that will be responsible for serving the files.

**Comment by dduportal** [comment]
Cooln thanks for the details.

What do you think of the following setup then:

- The "index" service is proxied by Traefik at the URL "http://index.yourcompany.org", and it's responsible for either generate the data in the response (if my understanding is correct), or answers a 302 redirect to URL "http://data.yourcompany.org/<id>".
- The "data" service is proxied by Traefik at the URL "http://data.yourcompany.org". A rewrite of the URL is done in order to map the backend storage system based on the initial request's URL.

Both of these services are configured with the middleware forward-auth:

- If a user tries to hit directly the storage system, it must be authenticated AND must know the exact URL of a given file.
- In normal case, the request emitted to the data service contains the right header generated after the authentication on the "index" service so no need to authenticate again.

**Comment by zyp** [comment]
I have already considered a combination of the ForwardAuth and ReplacePathRegex middlewares and two routers with different PathPrefix rules, which I figure would work if I'm careful, but it's not the way I would prefer to do it.

The middleware I propose would be easier to configure on the Traefik side and allow more flexibility.

**Comment by dduportal** [comment]
Hi @zyp , Thanks fro your feedbacks.

As for today, with Traefik 2.0 which entered the beta state last week, I recommend you to wait for the 2.0 release, when we'll have a better view on how the community uses middleware and asks for use cases.

We've put this issue in the category of proposal, as it is requesting an edge use case so we're not sure we'll implement this in the default list of middleware.

You are welcome to build such a middleware and compile Traefik with it if it fits your use case and can help!

**Comment by tirumerla** [comment]
Hi @dduportal, any update on this implementation. It would be great to have this middleware functionality.

My use case is i would want to redirect the client to another URL after hitting a backend service. Right now we don't have the flexibility of my backend service answering a standard Location HTTP header that would trigger the redirect at the client.

Any help would be appreciated.

**Comment by dduportal** [comment]
Hi @tirumerla , I'm not involved in this project anymore, but I'm sure @SantoDE  will take the the next steps there :)
Have a nice day!

**Comment by SantoDE** [comment]
Hello @tirumerla,

this is currently tracked as proposal, to see community need for that feature. Once we have more information or more visible demands, we might update that issue :)

**Comment by tirumerla** [comment]
> Hello @tirumerla,
> 
> this is currently tracked as proposal, to see community need for that feature. Once we have more information or more visible demands, we might update that issue :)

Thanks @SantoDE

**Comment by TheCapsLock** [comment]
I think the requested feature is similar to X-Accel-Redirect implemented by nginx : https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/
As an example, an app can use this mechanism to protect links to shared documents with a token (a dedicated service checks the token, if token is valid it answers with a X-Accel-Redirect which is internally processed by nginx which forwards that request handled by a private location in charge of serving the file. This mechanism prevents exposing the URL which gives access to documents without a proper authentification)

This kind of mechanism is - at least - used by [Taiga](https://www.taiga.io/)

**Comment by ddtmachado** [comment]
This seems to be a use case for pre-routing feature, which is being discussed on https://github.com/traefik/traefik/issues/5098 

If that's the case please leave feedback on the proposals for the configuration or anything else you feel is missing / could be improved.

**Comment by emilevauge** [comment]
Hello there,

This proposal has been open for a long time with minimal traction. We won't be adding it to our roadmap at this time and prefer to close it for now.

If circumstances change or interest grows, we can always reopen this later.
Thank you, @zyp , for the suggestion!
